### ISSUE TITLE
DNSCrypt and Unbound Solution

### ISSUE THREAD

**Issue by franciscomsilva** [notthebee/ansible-easy-vpn#118]
Hey,

Before explaining my sugestion, I would just like to appreciate for all of the work and dedication to your youtube channel and to this script specifically. It's really great and rather helpfull.

I am however facing some confusions regarding the configuration you chose. Let me try to explain it:

I'm not to familiar with DNSCrypt but I've had my headaches with Wireguard and Unbound for a few year now. 

As fair as I understand your DNS configuration is the following:

Adguard -> Unbound -> DNS Crypt -> Cloudflare/Google/Quad9/etc - > Root nameserver

Most implementation I have seen choose one of the following options:

- DNSCrypt and a public DNS recursive resolver, to encrypt the DNS trafic over HTTPS 
- Plain old Unbound which acts as a recursive DNS server and communicate directly with root nameservers, avoiding giving your DNS requests to a third-party

In my mind, the configuration you were aiming here should be:

Adguard -> DNS Crypt -> Unbound -> Root Nameserver

This would mean that the DNS requests would be encrypted and we would bypass the need to use some public recursive DNS such as Cloudflare and go straight to the root nameservers (using Unbound).

I believe this to be the most secure communication. On this implementation, using Unbound to then go to DNSCrypt and then something like Cloudflare, you are still using a third party DNS resolver and providing them with some information on your DNS requests, which in terms of privacy, might not be so good.

However, even the implementation I suggested raises some confusion: if you are using a DNS server inside our wireguard network and routing our DNS traffic to it, is not already encrypted? If the DNS server being used is 10.8.2.2, inside our PRIVATE wireguard network, are not all requests already being encrypted?  If my suspicious are correct, this means we could avoid usind DNSCrypt at all and just deploy AdGuard with Unbound.

A client sends a DNS request through the Wireguard tunnel to the Adguard container, runnning also inside the tunnel, which forwards it to Unbound and so one to the root nameserver. 

I have been trying to further understand the point of your configuration but, as I said, I am not too familiar with DNSCrypt, so, if I am at all mistaken, please let me know.

Sorry for the long post but kind of complex topics to be discussing.

Thank you again and have a nice year!

**Comment by notthebee** [comment]
Hey,

Mostly I just wanted to implement ad blocking via Adguard. But then I decided to also add DNS caching and DoH for a good measure 😁

The root nameservers way to do it is also definitely valid, and I will consider implementing it that way. At the same time I don't want to confuse the users with too much choice, so for now, i think the choice between CF, Google and Quad9 is the right way to go

**Comment by franciscomsilva** [comment]
Hey,

Thanks for the response. Valid option, I see your idea! 

Anyways, if in the future you want to implement something like I described it is rather easy, I did it with a few modification in no more than 30 minutes. But I agree that for now the choices you display are ok, specially since Quad9 also takes security and privacy a bit more serious than Google or Cloudflare. My option is for the ultimate privacy nerd :)

Anyways, thanks for the repplies and all of the work.

Have a nice week!
